File: CashuWallet/Core/Protocols/WalletServiceProtocol.swift
import Foundation
import CashuDevKit

// MARK: - Wallet Service Protocol

/// Protocol defining wallet operations.
/// This abstraction allows for different wallet implementations and facilitates testing.
protocol WalletServiceProtocol {
    /// Current total balance across all mints
    var balance: UInt64 { get async }
    
    /// Whether the wallet has been initialized
    var isInitialized: Bool { get }
    
    /// Initialize the wallet with a mnemonic
    func initialize(mnemonic: String) async throws
    
    /// Create a new wallet with a fresh mnemonic
    func createNewWallet() async throws -> String
    
    /// Restore wallet from mnemonic
    func restore(mnemonic: String) async throws
}

// MARK: - Mint Service Protocol

/// Protocol for mint management operations
protocol MintServiceProtocol {
    /// List of configured mints
    var mints: [MintInfo] { get async }
    
    /// Currently active mint
    var activeMint: MintInfo? { get async }
    
    /// Add a new mint
    func addMint(url: String) async throws -> MintInfo
    
    /// Remove a mint
    func removeMint(url: String) async throws
    
    /// Set the active mint
    func setActiveMint(_ mint: MintInfo) async throws
    
    /// Refresh mint info
    func refreshMintInfo(_ mint: MintInfo) async throws -> MintInfo
}

// MARK: - Token Service Protocol

/// Protocol for ecash token operations
protocol TokenServiceProtocol {
    /// Send tokens of specified amount
    func sendTokens(amount: UInt64, memo: String?, mintUrl: String?) async throws -> SendTokenResult
    
    /// Receive/redeem a token
    func receiveToken(_ tokenString: String) async throws -> UInt64
    
    /// Parse token information without redeeming
    func parseToken(_ tokenString: String) throws -> TokenInfo
    
    /// Check if a token has been spent
    func checkTokenSpent(_ tokenString: String) async throws -> Bool
}

// MARK: - Transaction Service Protocol

/// Protocol for transaction history management
protocol TransactionServiceProtocol {
    /// All transactions
    var transactions: [WalletTransaction] { get async }
    
    /// Load transactions from storage
    func loadTransactions() async throws
    
    /// Add a new transaction
    func addTransaction(_ transaction: WalletTransaction) async throws
    
    /// Update transaction status
    func updateTransactionStatus(id: String, status: WalletTransaction.TransactionStatus) async throws
}

// MARK: - Quote Service Protocol

/// Protocol for payment quote operations (mint/melt)
protocol QuoteServiceProtocol {
    /// Create a mint quote for the selected payment method.
    func createMintQuote(amount: UInt64?, method: PaymentMethodKind) async throws -> MintQuoteInfo
    
    /// Check mint quote status
    func checkMintQuote(id: String) async throws -> MintQuoteInfo

    /// Subscribe to quote updates when the mint and payment method support it.
    func subscribeToMintQuote(quoteId: String, paymentMethod: PaymentMethodKind) async throws -> ActiveSubscription?
    
    /// Mint tokens from a paid quote
    func mintTokens(quoteId: String) async throws -> UInt64
    
    /// Create a melt quote for a payment request.
    func createMeltQuote(request: String, preferredMintURL: String?) async throws -> MeltQuoteInfo
    
    /// Backward-compatible bolt11-specific entrypoint
    func createMeltQuote(invoice: String, preferredMintURL: String?) async throws -> MeltQuoteInfo

    /// Create a melt quote for an on-chain bitcoin address.
    func createOnchainMeltQuote(address: String, amount: UInt64, preferredMintURL: String?) async throws -> MeltQuoteInfo
    
    /// Execute the melt (pay the request)
    func meltTokens(quoteId: String, mintUrl: String?) async throws -> MeltPaymentResult
}

File: android/app/src/main/java/com/cashu/me/Core/Protocols/WalletServiceProtocol.kt
package com.cashu.me.Core.Protocols

import kotlinx.coroutines.flow.Flow
import com.cashu.me.Models.MeltPaymentResult
import com.cashu.me.Models.MeltQuoteInfo
import com.cashu.me.Models.MintInfo
import com.cashu.me.Models.MintQuoteInfo
import com.cashu.me.Models.PaymentMethodKind
import com.cashu.me.Models.RestoreMintResult
import com.cashu.me.Models.SendTokenResult
import com.cashu.me.Models.TokenInfo
import com.cashu.me.Models.TransactionStatus
import com.cashu.me.Models.WalletTransaction

interface WalletServiceProtocol {
    suspend fun initialize()
    suspend fun createNewWallet()
    suspend fun restoreWallet(mnemonic: String)
    suspend fun deleteWallet()
    suspend fun addMint(url: String)
    suspend fun removeMint(mint: MintInfo)
    suspend fun setActiveMint(mint: MintInfo)
    suspend fun restoreFromMint(url: String): RestoreMintResult
    suspend fun createMintQuote(amount: Long?, method: PaymentMethodKind, unit: String = "sat"): MintQuoteInfo
    suspend fun mintTokens(quoteId: String): Long
    suspend fun createMeltQuote(request: String, amountSats: Long? = null, preferredMintURL: String? = null): MeltQuoteInfo
    suspend fun meltTokens(quoteId: String, mintUrl: String? = null): MeltPaymentResult
    suspend fun sendTokens(amount: Long, memo: String?, p2pkPubkey: String?, mintUrl: String?, unit: String = "sat"): SendTokenResult
    suspend fun receiveTokens(tokenString: String): Long
}

interface MintServiceProtocol {
    suspend fun mints(): List<MintInfo>
    suspend fun activeMint(): MintInfo?
    suspend fun addMint(url: String): MintInfo
    suspend fun removeMint(url: String)
    suspend fun setActiveMint(mint: MintInfo)
    suspend fun refreshMintInfo(mint: MintInfo): MintInfo
}

interface TokenServiceProtocol {
    suspend fun sendTokens(amount: Long, memo: String? = null): SendTokenResult
    suspend fun receiveToken(tokenString: String): Long
    fun parseToken(tokenString: String): TokenInfo
    suspend fun checkTokenSpent(tokenString: String): Boolean
}

interface TransactionServiceProtocol {
    suspend fun transactions(): List<WalletTransaction>
    suspend fun loadTransactions()
    suspend fun addTransaction(transaction: WalletTransaction)
    suspend fun updateTransactionStatus(id: String, status: TransactionStatus)
}

interface QuoteServiceProtocol {
    suspend fun createMintQuote(amount: Long?, method: PaymentMethodKind): MintQuoteInfo
    suspend fun checkMintQuote(id: String): MintQuoteInfo
    fun subscribeToMintQuote(quoteId: String, paymentMethod: PaymentMethodKind): Flow<MintQuoteInfo>
    suspend fun mintTokens(quoteId: String): Long
    suspend fun createMeltQuote(request: String): MeltQuoteInfo
    suspend fun createBolt11MeltQuote(invoice: String): MeltQuoteInfo
    suspend fun createOnchainMeltQuote(address: String, amount: Long): MeltQuoteInfo
    suspend fun meltTokens(quoteId: String): MeltPaymentResult
}
